implementation.md

```diff
@@ -217,41 +215,88 @@ instance, storing it as a JSON file, then using that on another instance.
+and
+[`JiraConnection`](plugin/jira/src/main/java/ca/on/oicr/gsi/shesmu/jira/JiraConnection.java)
+or
+[`SftpDefinitionRepository`](plugin/sftp/src/main/java/ca/on/oicr/gsi/shesmu/sftp/SftpDefintionRepository.java)
+and
+[`SftpServer`](plugin/sftp/src/main/java/ca/on/oicr/gsi/shesmu/sftp/SftpServer.java).
+
+#### Constants
 A constant is a value that can be generated with no input. It need not actually
 be constant (_e.g._, `now` is a constant). The Shesmu compiler will arbitrarily
 copy the value of a constant, so a constant should be side-effect free.
 
-1. Create a class that implements `ConstantSource`.
-1. Annotate this class with `@MetaInfServices`.
-1. Include a no-arguments constructor for the `ConstantSource` class.
-1. Return a stream of `Constant` objects for the constants to be created.
-
-There are a number of convenience methods to create constants for values. If
-something more complicated is desired, extend the `Constant` class and write
-arbitrary bytecode in the `load` method. The code must not change any value on
-the stack and must leave exactly one value of the correct type on the stack.
-
-The `RuntimeBinding` class can be used to create constants as method calls on
-the instance of a class.
-
-### Functions
+##### Manual Implementation
+When directly implementing the `DefinitionRepository` interface, create
+instances of `ConstantDefinition`. There are a number of convenience
+methods to create constants for values. If something more complicated is
+desired, extend the `ConstantDefinition` class and write arbitrary bytecode in
+the `load` method. The code must not change any value on the stack and must
+leave exactly one value of the correct type on the stack.
+
+##### File-Backed Matched Implementation
+For file-backed matched implementations, constants are defined in the same way
+that functions are defined. Constants are simply functions that do not take any
+input arguments. See the section below for details.
+
+##### File-Backed Arbitrary Implementation
+For file-backed arbitrary implementations, constants can be created through the
+`UserDefiner.defineConstant` methods, which can take either a value or a
+function that returns that value.
+
+#### Functions
 A function is a transformation of input data. It matches with a call to a
 method. Since there is no user-defined error-handling, these functions should
 not throw. Also, since Shesmu has no null values, they should not return null
 when an error occurs.
 
-To provide functions:
-
-1. Create a class that implements `FunctionRepository`.
-1. Annotate this class with `@MetaInfServices`.
-1. Include a no-arguments constructor for the `FunctionRepository` class.
-1. Return a stream of `FunctionDefintion` objects for the functions to be created.
-
-There are a few ways methods to generate a `FunctionDefinition`:
+##### Manual Implementation
+When directly implementing the `DefinitionRepository` interface, create
```

Comment:
Same as above. Indicate why we would need to manually implement. With this section and the other manual implementation sections, I would suggest moving them after the File-Backed* sections: that way people can see if there is non-manual implementation that matches their use case before moving on to manual.